controllers/vspherevm_controller.go

```diff
@@ -0,0 +1,297 @@
+		// fetched resource has a a different ResourceVersion than the local
+		// object.
+		//
+		// FYI - resource versions are opaque, numeric strings and should not
+		// be compared with < or >, only for equality -
+		// https://kubernetes.io/docs/reference/using-api/api-concepts/#resource-versions.
+		//
+		// Since CAPV is currently deployed with a single replica, and this
+		// controller has a max concurrency of one, the only agent updating the
+		// VSphereVM resource should be this controller.
+		//
+		// So if the remote resource's ResourceVersion is different than the
+		// ResourceVersion of the resource fetched at the beginning of this
+		// reconcile request, then that means the remote resource should be
+		// newer than the local resource.
+		wait.PollImmediateInfinite(time.Second*1, func() (bool, error) {
+			// remoteObj refererences the same VSphereVM resource as it exists
+			// on the API server post the patch operation above. In a perfect world,
+			// the Status for localObj and remoteObj should be the same.
+			remoteObj := &infrav1.VSphereVM{}
+			if err := vmContext.Client.Get(vmContext, req.NamespacedName, remoteObj); err != nil {
+				if apierrors.IsNotFound(err) {
+					// It's possible that the remote resource cannot be found
+					// because it has been removed. Do not error, just exit.
+					return true, nil
+				}
+
+				// There was an issue getting the remote resource. Sleep for a
+				// second and try again.
+				vmContext.Logger.Error(err, "failed to get VSphereVM while exiting reconcile")
+				return false, nil
+			}
+
+			// If the remote resource version is not the same as the local
+			// resource version, then it means we were able to get a resource
+			// newer than the one we already had.
+			if localObj.ResourceVersion != remoteObj.ResourceVersion {
+				vmContext.Logger.Info(
+					"resource is patched",
+					"local-resource-version", localObj.ResourceVersion,
+					"remote-resource-version", remoteObj.ResourceVersion)
+				return true, nil
+			}
+
+			// If the resources are the same resource version, then a previous
+			// patch may not have resulted in any changes. Check to see if the
+			// remote status is the same as the local status.
+			if cmp.Equal(localObj.Status, remoteObj.Status, cmpopts.EquateEmpty()) {
+				vmContext.Logger.Info(
+					"resource patch was not required",
+					"local-resource-version", localObj.ResourceVersion,
+					"remote-resource-version", remoteObj.ResourceVersion)
+				return true, nil
+			}
+
+			// The remote resource version is the same as the local resource
+			// version, which means the local cache is not yet up-to-date.
+			vmContext.Logger.Info(
+				"resource is not patched",
+				"local-resource-version", localObj.ResourceVersion,
+				"remote-resource-version", remoteObj.ResourceVersion)
+			return false, nil
+		})
+	}()
+
+	// Handle deleted machines
+	if !vsphereVM.ObjectMeta.DeletionTimestamp.IsZero() {
+		return r.reconcileDelete(vmContext)
```

Comment:
nit: reconcileDelete -> reconcileDeletion

Reply:
Hi @tedyu,

I think that's right? Regardless, this pattern is used across all CAPI providers, so I'll likely leave this alone since it's the same in CAPI, CAPA, CAPG, CAPD, etc.

Reply:
Alright.
I was not aware of the other places where this name is used.